### PR TITLE
telink: b92: fix partial flash protect .

### PR DESCRIPTION
**Pull request by haiwentelink**
- change the flash protect to adpot partial protect.

### CODE REVIEW

**Comment by andriy-bilynskyy** [comment]
The following west manifest projects have been modified in this Pull Request:

| Name | Old Revision | New Revision | Diff |
| ---- | ------------ | ------------ |------|
| hal_telink | https://github.com/telink-semi/hal_telink/commit/8e22fad3c938f4ea2266293897d8b8ea47e566a0 | https://github.com/telink-semi/hal_telink/commit/34700f71d3c69326595484c631853ca61fabf0af (develop) | [telink-semi/hal_telink@8e22fad3..34700f71](https://github.com/telink-semi/hal_telink/compare/8e22fad3c938f4ea2266293897d8b8ea47e566a0..34700f71d3c69326595484c631853ca61fabf0af) |

*Note: This message is automatically posted and updated by the Manifest GitHub Action.*

File: drivers/flash/soc_flash_b9x.c [hunk 1]
```diff
@@ -105,31 +117,39 @@ static void flash_b9x_modify(struct flash_b9x_data *dev_data, uintptr_t offset,
...
 			len_current = len;
 		}
-		flash_read_page(addr_flash, len_current, &dev_data->sector[off_sector]);
-		if (!flash_b9x_is_clean(dev_data, off_sector, len_current)) {
+		clean_flag = flash_b9x_is_clean(dev_data, addr_flash, off_sector, len_current);
+		if (!clean_flag) {
+			uint8_t *p_sector = malloc(SECTOR_SIZE);
+
+			if (p_sector == NULL) {
+				return;
```
Comment (andriy-bilynskyy) [inline drivers/flash/soc_flash_b9x.c:118]: Please add LOG_ERR here

File: drivers/flash/soc_flash_b9x.c [hunk 2]
```diff
@@ -105,31 +117,39 @@ static void flash_b9x_modify(struct flash_b9x_data *dev_data, uintptr_t offset,
 	while (len) {
 		size_t len_page_end =  SECTOR_SIZE - off_sector;
 		size_t len_current = len_page_end;
+		bool clean_flag = true;
```
Comment (andriy-bilynskyy) [inline drivers/flash/soc_flash_b9x.c]: No needed to assign this value.

File: drivers/flash/soc_flash_b9x.c [hunk 3]
```diff
@@ -37,22 +37,24 @@ static const struct flash_parameters flash_b9x_parameters = {
...
-	flash_protection_unlock_operation();
+	flash_protection_unlock_operation(offset);
 #elif CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B95
 
 #endif
 
 }
 
-static void flash_b9x_lock_start(void)
+static unsigned int lock_offset;
```
Comment (andriy-bilynskyy) [inline drivers/flash/soc_flash_b9x.c]: Static variableas are not allowed in drivers. Each device has own data and configuration.

Comment (andriy-bilynskyy) [inline drivers/flash/soc_flash_b9x.c]: What this offset is ding?

File: drivers/flash/soc_flash_b9x.c [hunk 4]
```diff
@@ -346,17 +366,9 @@ static const struct flash_driver_api flash_b9x_api = {
...
-	__attribute__((section(".bss")))			    \
-	static uint8_t sector_##n[SECTOR_SIZE];			    \
-										    \
-	static struct flash_b9x_data flash_data_##n = {			    \
-		.sector = sector_##n			    \
-	};			    \
-										    \
 	DEVICE_DT_INST_DEFINE(n, flash_b9x_init,			\
 			      NULL,				    \
-			      &flash_data_##n,				    \
```
Comment (andriy-bilynskyy) [inline drivers/flash/soc_flash_b9x.c:359]: That's Zephyr drivers convention to use device data not static variables.
It solves the problem with multiply instances.

File: drivers/flash/soc_flash_b9x.c [hunk 5]
```diff
@@ -15,6 +15,7 @@
 #include <zephyr/device.h>
 #include <zephyr/drivers/flash.h>
 #include <zephyr/kernel.h>
+#include <stdlib.h>
```
Comment (andriy-bilynskyy) [inline drivers/flash/soc_flash_b9x.c:19]: The driver is already using k_malloc which allocates data in system heap not user.
For now it's using both heaps.
Clear way is to use only system heap for drivers. But if we want to have memory pofit lets use user in all driver code, anyway we have no MPU existed/setuped.

**Review by andriy-bilynskyy — CHANGES_REQUESTED** [review]

File: drivers/flash/soc_flash_b9x.c (hunk 5 above)
Comment (haiwentelink) [reply drivers/flash/soc_flash_b9x.c:19]: the init config heap is configed to 512, but the heap of the lib is very large(12k) , so I think there is no need to enlarge the system heap , multi use of the heap .

File: drivers/flash/soc_flash_b9x.c (hunk 4 above)
Comment (haiwentelink) [reply drivers/flash/soc_flash_b9x.c:359]: if don't use the flash_data, I think this value can set to be null .

File: drivers/flash/soc_flash_b9x.c (hunk 3 above)
Comment (haiwentelink) [reply drivers/flash/soc_flash_b9x.c]: we want to transfer the offset address to lock func , but the lock fun will not run immidately ,if not keep , the value will lost .

File: drivers/flash/soc_flash_b9x.c (hunk 1 above)
Comment (haiwentelink) [reply drivers/flash/soc_flash_b9x.c:118]: will add in this func .

File: drivers/flash/soc_flash_b9x.c (hunk 2 above)
Comment (haiwentelink) [reply drivers/flash/soc_flash_b9x.c]: will change

File: drivers/flash/soc_flash_b9x.c (hunk 3 above)
Comment (haiwentelink) [reply drivers/flash/soc_flash_b9x.c]: can we assign the adr into the struct of the flash_b9x_parameters

**Comment by haiwentelink** [comment]
@andriy-bilynskyy  can start review again , delete the timer for lock and unlock , can call directly , for it will not excute frequently . and fix the bugs of the pwm driver for the interface changed .

**Review by andriy-bilynskyy — APPROVED** [review]

**Comment by andriy-bilynskyy** [comment]
Plese update HAL Telink 34700f71d3c69326595484c631853ca61fabf0af

**Pull request merged**